src/Components/WebAssembly/Server/src/DebugProxyLauncher.cs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    internal static class DebugProxyLauncher
+    {
+        private static readonly object LaunchLock = new object();
+        private static readonly TimeSpan DebugProxyLaunchTimeout = TimeSpan.FromSeconds(10);
+        private static Task<string> LaunchedDebugProxyUrl;
+        private static readonly Regex NowListeningRegex = new Regex(@"^\s*Now listening on: (?<url>.*)$", RegexOptions.None, TimeSpan.FromSeconds(10));
+        private static readonly Regex ApplicationStartedRegex = new Regex(@"^\s*Application started\. Press Ctrl\+C to shut down\.$", RegexOptions.None, TimeSpan.FromSeconds(10));
+
+        public static Task<string> EnsureLaunchedAndGetUrl(IServiceProvider serviceProvider)
+        {
+            lock (LaunchLock)
+            {
+                if (LaunchedDebugProxyUrl == null)
+                {
+                    LaunchedDebugProxyUrl = LaunchAndGetUrl(serviceProvider);
+                }
+
+                return LaunchedDebugProxyUrl;
+            }
+        }
+
+        private static async Task<string> LaunchAndGetUrl(IServiceProvider serviceProvider)
```

Comment:
Can I suggest that instead or relying on the Regexes (given that we control the proxy) we define on the server a "ready" endpoint and pass that down to the proxy and have the proxy send an Http request back to the server to indicate it's up and running? That way we avoid having to depend on regexes and reading console output and I think we would have a more reliable solution.

We can actually do this for keeping the process alive too, by passing in a "keep alive" endpoint that the proxy can ping to determine if the launcher process is still alive and terminate itself after a defined period of time if that's not the case. That way you don't just rely on the parent PID (which although is very rare, can be reused sometimes)

Reply:
I'd like to not do that unless it becomes necessary. That would be yet more moving parts and a more complex polling system.

The approach of regexing console output is pretty commonplace for this purpose. We did it for NodeServices where it's continued to work fine for many years, plus we do it in our tests. VS Code does the same thing too when launching ASP.NET Core and connecting the debugger.

We don't need to keep the process alive. It stays alive on its own. We're using the parent's only when starting up the listener, and thereafter we attach to the `Exited` event which is designed for this scenario.

If any of this turns out to be inadequate in real use I'm very open to changing it, but current have no reason to think there are realistic drawbacks to this approach.